### PR TITLE
✨ Remade EUDEMO TF coil

### PR DESCRIPTION
**Pull request by je-cook**
## Description

Done from my end 

## Checklist

I confirm that I have completed the following checks:

- [X] Tests run locally and pass `pytest tests --reactor`
- [X] Code quality checks run locally and pass `flake8` and `black .`
- [X] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/builders/tools.py [hunk 1]
```diff
@@ -85,7 +85,7 @@ def get_n_sectors(no_obj, degree=360):
 
 
 def circular_pattern_component(
-    component: bm_comp.Component,
+    component: Union[bm_comp.Component, List[bm_comp.Component]],
```
Comment (je-cook) [inline bluemira/builders/tools.py:88]: This rearrangement means you can have multiple components all under the same sector patterned without ending up with an extra level

File: examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        "tf_wp_depth": {"value": 1.00},
+        "tk_tf_front_ib": {"value": 0.04},
+        "tk_tf_ins": {"value": 0.08},
+        "tk_tf_insgap": {"value": 0.1},
+        "tk_tf_nose": {"value": 0.6},
+        "tk_tf_side": {"value": 0.1},
+    }
+
+    @pytest.mark.parametrize("centreline, wp_xs", zip(centrelines, wp_xs))
+    def test_components_and_segments(self, centreline, wp_xs):
```
Comment (je-cook) [inline examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py:215]: these two test takes a while (21s locally) considering long running them, thoughts?

File: examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py [hunk 3]
```diff
@@ -0,0 +1,805 @@
...
+        shape_params = {}
+        for key, val in variables_map.items():
+            if isinstance(val, str):
+                val = getattr(self.params, val).value
+
+            if isinstance(val, dict):
+                if isinstance(val["value"], str):
+                    val["value"] = getattr(self.params, val["value"]).value
+            else:
+                val = {"value": val}
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py:216]: Is `variables_map` being mutated here? Should it be?

File: examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py [hunk 4]
```diff
@@ -0,0 +1,805 @@
...
+        """
+        Run the specified design optimisation problem to generate the TF coil winding
+        pack current centreline.
+        """
+        parameterisation = self._get_parameterisation()
+        wp_cross_section = self._make_wp_xs(self.params.r_tf_current_ib.value)
+
+        if not hasattr(self, "problem_class"):
+            raise ValueError(
+                f"Cannot execute {type(self).__name__} in RUN mode: no problem_class specified."
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py:266]: We have a similar error on line 318, but `read` is in lowercase, where we have uppercase here. My personal preference is lower case (as that's the function name), but I'm happy with either, as long as we're consistent.

File: examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py [hunk 5]
```diff
@@ -0,0 +1,805 @@
...
+            else self._make_centreline_koz(self.keep_out_zone),
+            **self.problem_settings,
+        )
+
+        bluemira_print(f"Solving design problem: {type(design_problem).__name__}")
+        if parameterisation.n_ineq_constraints > 0:
+            bluemira_debug("Applying shape constraints")
+            design_problem.apply_shape_constraints()
+
+        bluemira_debug("Solving...")
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py:308]: We're logging that we're '`Solving`' twice here

File: examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
+# bluemira is an integrated inter-disciplinary design tool for future fusion
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py:1]: Very minor, but the source file is `tf_coils`, so to match up, this file should probably be called `test_tf_coils`

File: examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+                },
+            }
+        )
+        designer = TFCoilDesigner(
+            self.PARAMS,
+            build_config=config,
+            separatrix=self.lcfs,
+            keep_out_zone=self.vvts_koz,
+        )
+        d_run, wp_run = designer.execute()
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py:147]: `wp_run` isn't used
```suggestion
        d_run, _ = designer_mock.execute()
```

File: examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+        designer_mock = TFCoilDesigner(
+            self.PARAMS,
+            build_config=self.CONFIG,
+            separatrix=self.lcfs,
+            keep_out_zone=self.vvts_koz,
+        )
+        d_mock, wp_run = designer_mock.execute()
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py:155]: Ditto
```suggestion
        d_mock, _ = designer_mock.execute()
```

**Review by hsaunders1904 — CHANGES_REQUESTED** [review]
This is looking good, a few suggestions

File: examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py (hunk 7 above)
Comment (hsaunders1904) [reply examples/design/EU-DEMO/EUDEMO_tests/test_tfcoils.py]: Whoops, this should've been `designer.execute`, my bad

File: examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py (hunk 3 above)
Comment (je-cook) [reply examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py:214]: Unsure, changed it so it definitely isnt mutated

File: examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py [hunk 9]
```diff
@@ -0,0 +1,828 @@
...
+class TFCoilDesigner(Designer[GeometryParameterisation]):
+    """
+    TF Coil Designer
+
+    Parameters
+    ----------
+    params: Union[Dict, ParameterFrame]
+        TF Coil Designer parameters
+    build_config: Dict
+        Required keys;
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py:144]:
```suggestion
        Required keys:
```

File: examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py [hunk 10]
```diff
@@ -0,0 +1,828 @@
...
+
+            * param_class: str
+                A string of the import location for the parameterisation
+                class of the TF Coil
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py:148]: An example could be useful here, as it's not immediately obvious to me what the syntax should be. Something like: "e.g., `bluemira.geometry.parameterisations::PictureFrame`"?

File: examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py [hunk 11]
```diff
@@ -0,0 +1,828 @@
...
+
+        Optional keys:
+
+            * variables_map: Dict
+                variable map for the param_class
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py:153]: Is it worth clarifying what exactly is being mapped? Is it parameter names to `param_class` constructor kwargs?

File: examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py [hunk 12]
```diff
@@ -0,0 +1,828 @@
...
+
+        joiner_top, joiner_bottom = self._make_joining_sections(
+            z_min, z_max, centreline_points
+        )
+
+        # Cut away straight sweep before fusing to protect against degenerate faces
+        # Keep the largest piece
+        pieces = boolean_cut(solid, inboard_casing)
+        pieces.sort(key=lambda x: x.volume)
+        solid = pieces[-1]
```
Comment (hsaunders1904) [inline examples/design/EU-DEMO/EUDEMO_builders/tf_coils.py:645]:
```suggestion
        solid = max(pieces, key=lambda x: x.volume)
```

**Review by hsaunders1904 — COMMENTED** [review]
A few nitpicks, but this looks good. 

I'm still getting a crash when I try to interact with the CAD view generated in the tests, though (well, it crashes on interaction the first time I run the test, after that it hangs and then crashes before the CAD viewer is opened). 

<details><summary>Stacktrace</summary>
<p>

```console
pytest examples/design/EU-DEMO/EUDEMO_tests/test_tf_coils.py --plotting-on
======================================================= test session starts ========================================================
platform linux -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/bf2936/bluemira/code/bluemira, configfile: pyproject.toml
plugins: forked-1.4.0, metadata-2.0.2, cov-3.0.0, json-report-1.5.0, repeat-0.9.1, xdist-2.5.0, html-3.1.1, typeguard-2.13.3
collected 7 items

examples/design/EU-DEMO/EUDEMO_tests/test_tf_coils.py .....Fatal Python error: Aborted

Current thread 0x00007efd20dfc740 (most recent call first):
  File "/home/bf2936/bluemira/code/bluemira/bluemira/codes/_freecadapi.py", line 1724 in show_cad
  File "/home/bf2936/bluemira/code/bluemira/bluemira/display/displayer.py", line 172 in show_cad
  File "/home/bf2936/bluemira/code/bluemira/bluemira/display/displayer.py", line 218 in show_cad
  File "/home/bf2936/bluemira/code/bluemira/bluemira/display/displayer.py", line 268 in show_cad
  File "/home/bf2936/bluemira/code/bluemira/examples/design/EU-DEMO/EUDEMO_tests/test_tf_coils.py", line 229 in test_components_and_segments
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/python.py", line 192 in pytest_pyfunc_call
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/python.py", line 1761 in runtest
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/runner.py", line 166 in pytest_runtest_call
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/runner.py", line 259 in <lambda>
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/runner.py", line 338 in from_call
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/runner.py", line 258 in call_runtest_hook
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/runner.py", line 219 in call_and_report
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/runner.py", line 130 in runtestprotocol
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/runner.py", line 111 in pytest_runtest_protocol
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/main.py", line 347 in pytest_runtestloop
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/main.py", line 322 in _main
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/main.py", line 268 in wrap_session
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/main.py", line 315 in pytest_cmdline_main
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_callers.py", line 39 in _multicall
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_manager.py", line 80 in _hookexec
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/pluggy/_hooks.py", line 265 in __call__
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/config/__init__.py", line 164 in main
  File "/home/bf2936/miniforge3/envs/bluemira/lib/python3.8/site-packages/_pytest/config/__init__.py", line 187 in console_main
  File "/home/bf2936/miniforge3/envs/bluemira/bin/pytest", line 8 in <module>
[2]    30600 abort      pytest examples/design/EU-DEMO/EUDEMO_tests/test_tf_coils.py --plotting-on
```

</p>
</details>

**Comment by hsaunders1904** [comment]
>  it crashes on interaction the first time I run the test, after that it hangs and then crashes before the CAD viewer is opened

Figured out what's actually happening. Interacting with the viewer crashes XLaunch, so that's why it was hanging when I ran it afterwards: no display server was running. So it could potentially be an XLaunch issue 😞

**Pull request merged**